File: projects/java/cdm/src/mil/tatrc/physiology/datamodel/substance/SESubstanceFraction.java
/**************************************************************************************
Copyright 2015 Applied Research Associates, Inc.
Licensed under the Apache License, Version 2.0 (the "License"); you may not use
this file except in compliance with the License. You may obtain a copy of the License
at:
http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
CONDITIONS OF ANY KIND, either express or implied. See the License for the
specific language governing permissions and limitations under the License.
**************************************************************************************/

package mil.tatrc.physiology.datamodel.substance;

import mil.tatrc.physiology.datamodel.CDMSerializer;
import mil.tatrc.physiology.datamodel.bind.*;
import mil.tatrc.physiology.datamodel.properties.*;

public class SESubstanceFraction
{
	protected SEScalarFraction  fractionAmount;
	protected SESubstance       substance;
	
	public SESubstanceFraction(SESubstance s)
	{
		if(s==null)
			throw new RuntimeException("Must provide a valid substance");
		this.substance=s;
	}
	
	public void reset()
	{
		if(this.fractionAmount!=null)
			this.fractionAmount.invalidate();
	}
	
	public void copy(SESubstanceFraction from)
	{
	  this.substance=from.substance;
	  if(from.hasFractionAmount())
	    this.getFractionAmount().set(from.getFractionAmount());
	}
	
	public boolean load(SubstanceFractionData data)
	{
		this.reset();
		if(data.getFractionAmount()!=null)
			this.getFractionAmount().load(data.getFractionAmount());
		return true;
	}
	
	public SubstanceFractionData unload()
	{
		SubstanceFractionData to = CDMSerializer.objFactory.createSubstanceFractionData();
		unload(to);
		return to;
	}
	
	protected void unload(SubstanceFractionData to)
	{
		if(hasFractionAmount())
			to.setFractionAmount(this.fractionAmount.unload());		
		to.setName(this.substance.getName());
	}
	
	public SEScalarFraction getFractionAmount() 
	{ 
		if(this.fractionAmount==null)
			this.fractionAmount=new SEScalarFraction();
		return this.fractionAmount;
	}
	public boolean hasFractionAmount() {return this.fractionAmount==null?false:this.fractionAmount.isValid();}
	
	public boolean hasSubstance() { return substance != null; }
	public SESubstance getSubstance() 
	{ 
		return this.substance;
	}
	
	public String toString()
	{
		if(!hasSubstance())
			return "";
		String str = "Fraction of "+this.substance.getName()+": "+(hasFractionAmount()?getFractionAmount():"None");			
		return str;
	}
}

File: projects/biogears/libBiogears/src/cdm/substance/SESubstanceFraction.cpp
/**************************************************************************************
Copyright 2015 Applied Research Associates, Inc.
Licensed under the Apache License, Version 2.0 (the "License"); you may not use
this file except in compliance with the License. You may obtain a copy of the License
at:
http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
CONDITIONS OF ANY KIND, either express or implied. See the License for the
specific language governing permissions and limitations under the License.
**************************************************************************************/
#include <biogears/cdm/substance/SESubstanceFraction.h>

#include "io/cdm/Substance.h"

#include <biogears/cdm/properties/SEScalarFraction.h>
#include <biogears/cdm/substance/SESubstance.h>
#include <biogears/cdm/substance/SESubstanceManager.h>

namespace biogears {
SESubstanceFraction::SESubstanceFraction(SESubstance const& substance)
  : m_Substance(substance)
{
  m_FractionAmount = nullptr;
}
SESubstanceFraction::SESubstanceFraction(SESubstance const& substance, SEScalarFraction const& fraction)
  : m_Substance(substance)
  , m_FractionAmount(nullptr)
{
  GetFractionAmount() = fraction;
}
//-------------------------------------------------------------------------------
SESubstanceFraction::~SESubstanceFraction()
{
  Clear();
}
//-------------------------------------------------------------------------------
void SESubstanceFraction::Clear()
{
  SAFE_DELETE(m_FractionAmount);
}
//-------------------------------------------------------------------------------
bool SESubstanceFraction::HasFractionAmount() const
{
  return m_FractionAmount == nullptr ? false : m_FractionAmount->IsValid();
}
//-------------------------------------------------------------------------------
SEScalarFraction& SESubstanceFraction::GetFractionAmount()
{
  if (m_FractionAmount == nullptr)
    m_FractionAmount = new SEScalarFraction();
  return *m_FractionAmount;
}
//-------------------------------------------------------------------------------
double SESubstanceFraction::GetFractionAmount() const
{
  if (m_FractionAmount == nullptr)
    SEScalar::dNaN();
  return m_FractionAmount->GetValue();
}
//-------------------------------------------------------------------------------
SESubstance& SESubstanceFraction::GetSubstance() const
{
  return const_cast<SESubstance&>(m_Substance);
}
//-------------------------------------------------------------------------------
bool SESubstanceFraction::operator==(const SESubstanceFraction& rhs) const
{
  bool equivilant = m_Substance == rhs.m_Substance;
  equivilant &= (m_FractionAmount && rhs.m_FractionAmount) ? m_FractionAmount->operator==(*rhs.m_FractionAmount) : m_FractionAmount == rhs.m_FractionAmount;
  return equivilant;
}
//-------------------------------------------------------------------------------
bool SESubstanceFraction::operator!=(const SESubstanceFraction& rhs) const
{
  return !(*this == rhs);
}
}